### PR TITLE
111799/project filters performance

### PR DESCRIPTION
**Pull request by tim-wilde-nimble**
This 'fixes' the performance problem by batching the requests for groups and trusts when loading projects. Though this will still experience performance degradation linearly in line with the number of projects in the database, it is at least down from several minutes to sub 1 second with the current test data set of approximately 1100 projects.

### CODE REVIEW

File: TramsDataApi/UseCases/SearchAcademyTransferProjects.cs [hunk 1]
```diff
@@ -3,48 +3,103 @@
...
+      {
+         IEnumerable<AcademyTransferProjects> academyTransferProjects = FilterByUrn(
+            _academyTransferProjectGateway.GetAcademyTransferProjects(), urn).ToList();
+
+         IEnumerable<AcademyTransferProjectSummaryResponse> projects =
+            FilterByIncomingTrust(title, AcademyTransferProjectSummaryResponse(academyTransferProjects));
+
+         var recordTotal = projects.Count();
+         projects = projects.OrderByDescending(atp => atp.ProjectUrn)
+            .Skip((page - 1) * 10).Take(10).ToList();
```
Comment (jordan-beard) [inline TramsDataApi/UseCases/SearchAcademyTransferProjects.cs:35]: Should the '10' in this query be replaced with the count?

File: TramsDataApi/UseCases/SearchAcademyTransferProjects.cs [hunk 2]
```diff
@@ -3,48 +3,103 @@
...
+
+         return Task.FromResult(new PagedResult<AcademyTransferProjectSummaryResponse>(projects, recordTotal));
+      }
+
+      private static IEnumerable<AcademyTransferProjects> FilterByUrn(IEnumerable<AcademyTransferProjects> queryable,
```
Comment (jordan-beard) [inline TramsDataApi/UseCases/SearchAcademyTransferProjects.cs:40]: Just a thought - you could return this as a dictionary. I think Linq offers a ToDictionary extension.

File: TramsDataApi/UseCases/SearchAcademyTransferProjects.cs (hunk 1 above)
Comment (tim-wilde-nimble) [reply TramsDataApi/UseCases/SearchAcademyTransferProjects.cs:35]: using the `count`parameter.

File: TramsDataApi/Gateways/TrustGateway.cs [hunk 3]
```diff
@@ -1,69 +1,81 @@
...
-    }
-}
\ No newline at end of file
+         return (
+            filteredGroups.Skip((page - 1) * count).Take(count).ToList(),
+            filteredGroups.Count()
+         );
+      }
+
+      public IEnumerable<Group> GetMultipleGroupsByUkprn(IEnumerable<string> ukprns)
```
[the remark this answers is not in the document]
Comment (tim-wilde-nimble) [reply TramsDataApi/Gateways/TrustGateway.cs:69]: Fair point. I just prefer the difference between this and `GetGroupByUkprn` to be a bit more visible than one character, so I put in 'Multiple' in both variants of methods that return more than one instance - to be more discernible at a glance.

File: TramsDataApi/UseCases/SearchAcademyTransferProjects.cs (hunk 2 above)
Comment (jordan-beard) [reply TramsDataApi/UseCases/SearchAcademyTransferProjects.cs:40]: Fair enough.

File: TramsDataApi/Gateways/TrustGateway.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jordan-beard) [reply TramsDataApi/Gateways/TrustGateway.cs:69]: Fair enough.

**Review by jordan-beard — APPROVED** [review]

**Pull request merged**